### PR TITLE
Parallelize per-contract storage updates

### PR DESCRIPTION
**Pull request by omerfirmak**
Substantial speed ups

before 

```
omer@omer-ThinkPad-E14-Gen-3:~/Documents/juno$ go test ./... -bench=MainnetBlock -benchmem
PASS
ok      github.com/NethermindEth/juno/adapters/feeder2core      0.402s
goos: linux
goarch: amd64
pkg: github.com/NethermindEth/juno/blockchain
cpu: AMD Ryzen 7 5700U with Radeon Graphics         
BenchmarkMainnetBlock-16           10         917313646 ns/op        18574441 B/op     434711 allocs/op
```

after

```
omer@omer-ThinkPad-E14-Gen-3:~/Documents/juno$ go test ./... -bench=MainnetBlock -benchmem
PASS
ok      github.com/NethermindEth/juno/adapters/feeder2core      0.385s
goos: linux
goarch: amd64
pkg: github.com/NethermindEth/juno/blockchain
cpu: AMD Ryzen 7 5700U with Radeon Graphics         
BenchmarkMainnetBlock-16           10         544530601 ns/op        21892010 B/op     464761 allocs/op
PASS
```

we can optimize for memory later

### CODE REVIEW

File: db/buffered_transaction.go
Comment (joshklop) [inline db/buffered_transaction.go:63]: nit: we can test `value == nil` and swap the branches.

**Review by joshklop — COMMENTED** [review]
Amazing. I am so jealous I did not think of this :)

File: core/state.go
Comment (omerfirmak) [reply core/state.go:370]: > unnecessary branching

I think it looks better if it is in the callback. Also, that branch is almost entirely static, so it will always be correctly predicted by the processor. So it has near-0 performance impact.

File: core/state.go
Comment (omerfirmak) [reply core/state.go:434]: Nope, because we shouldnt modify `s.txn` before we make sure that all contract storages are updated (ie `contractUpdaters.Wait()` returns). `s.txn` is essentially a `pebble.Batch` and it is not safe to be modified concurrently or modified while another routine is reading from it.

Think of it as a resource that needs to be protected with a `RWMutex`. But by delaying any writes to the transaction until we are done with updating contract storages by using a `BufferedTransaction`, we avoid needing a mutex.

Comment (joshklop) [reply core/state.go:434]: Can we continue making the updates in the buffered transaction?

File: db/buffered_transaction.go
Comment (omerfirmak) [reply db/buffered_transaction.go:63]: Done

File: core/state.go
Comment (omerfirmak) [reply core/state.go:434]: During the concurrent contract storage updates?

Yes, because it never writes to the transaction until `Flush()` or `Commit()` is called explicitly. It just keeps track of what changes were made on the `BufferedTransaction` in memory. Essentially it allows our goroutines to access `s.txn` in read-only mode, without them knowing it.

Later on, we apply those changes to `s.txn` all at the same time.

Comment (omerfirmak) [reply core/state.go:434]: https://github.com/NethermindEth/juno/pull/900/files#diff-c075eb780b39817edd44ca2456cee339e3c84f432658a27a58cbb44bc9204780R428-R432

Actually if we sort the slice of `BufferedTransaction`s we get back from the routines in ascending contract address order, we might get an additional speed up as well. But it is not really worth it at this point. Will leave it to another PR.

File: core/state.go
Comment (omerfirmak) [inline core/state.go:432]: Actually if we sort the slice of BufferedTransactions we get back from the routines in ascending contract address order, we might get an additional speed up. But it is not really worth it at this point. Will leave it to another PR.

File: core/state.go
Comment (omerfirmak) [reply core/state.go:434]: at L435?

No, we can't. because `updateContractCommitment` operates on the state trie regardless of which contract it was called with. All operations on a single Trie instance need to be done sequentially

Comment (joshklop) [reply core/state.go:434]: One more related question: why not just use an `RWMutex` instead of creating a new transaction implementation?

Comment (omerfirmak) [reply core/state.go:434]: I expected a lock to have a worse performance impact compared to memory allocations that we need for buffering.

But that is just my gut feeling. We can keep experimenting after merging this.

But regardless, we will have to implement a new type that wraps a transaction and protects it with a RWMutex.

**Review by joshklop — APPROVED** [review]

File: core/state.go
Comment (joshklop) [reply core/state.go:434]: Do you mind creating issues for this Mutex thing and the sorting idea? I can if you don't have time.

**Pull request merged**
